Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,242 @@
+/-
+Copyright (c) 2024 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero, Laura Capuano, Amos Turchet
+-/
+import Mathlib.Analysis.Matrix
+import Mathlib.Data.Pi.Interval
+
+/-!
+# Siegel's Lemma
+
+In this file we introduce and prove Siegel's Lemma in its most basic version. This is a fundamental
+tool in diophantine approximation and transcendency and says that there exists a "small" integral
+non-zero solution of a non-trivial underdetermined system of linear equations with integer
+coefficients.
+
+## Main results
+
+- `exists_ne_zero_int_vec_norm_le`: Given a non-singular `m × n` matrix `A` with `m < n` the linear
+system it determines has a non-zero integer solution `t` with
+`‖t‖ ≤ ((n * ‖A‖) ^ ((m : ℝ) / (n - m)))`
+
+## Notation
+
+ - `‖_‖ ` : Matrix.seminormedAddCommGroup is the sup norm, the maximum of the absolute values of
+ the entries of the matrix
+
+## References
+
+See [M. Hindry and J. Silverman, Diophantine Geometry: an Introduction][hindrysilverman00] .
+-/
+
+/- We set ‖⬝‖ to be Matrix.seminormedAddCommGroup  -/
+attribute [local instance] Matrix.seminormedAddCommGroup
+
+namespace Matrix
+
+open Matrix Finset
+
+variable {m n : Type*} [Fintype m] [Fintype n]
+
+/-- The norm of an integral matrix as the sup of the sup of the natAbs of the entries -/
+def natNorm (A : Matrix m n ℤ) : ℕ := sup univ fun i ↦ sup univ fun j ↦ (A i j).natAbs
```

Comment:
Is it really too painful to use the "standard" norm?

Reply:
@faenuccio.. People seem not to like this `natNorm`.
After some thought I think I can get rid of it, and work with $\|| A \||$ only. It is going to be quite a bit of work and I think the code is going to look uglier, or maybe not... who knows...
What do you say? Should I try to do it?

Reply:
The "normal" norm is a real number?

Reply:
Yes

Reply:
I think we will have the good answer only after some trials. In case you end up duplicating a lot of code (especially to prove that this norm is actually integral/positive) then it will be worth it create something, but this will become clearer only after some trial-and-error. You can start without and see how things go.

Reply:
Let me see if I can avoid it.

Reply:
I think one can make a lemma that extracts the indices `i` and `j` that realise the sups in the norm and then work with that. It would be something like $\exists i, j :  \|| A \|| = \|| A_{ij}\||$. The RHS is by definition the cast of a `natAbs (A i j)`.

Reply:
The real question is if you think this is a useful notion for others results (I guess it is?): in that case it is worth to create a new file, write the def and develop a basic API (even just the lemma you already have).

Reply:
Let me think about it. I'll try a couple of things